horace_core/sqw/file_io/class_helpers/pix_cache.m

```diff
@@ -106,74 +227,114 @@
-                    obj.read_pix_cache_{i}  = {};
+                    right = edges(2:end)-1;
+                    last_bin_num = nbin(end);
+                    if obj.last_bin_completed_(i)
+                        obj.bin_range_(1,i)    = last_bin_num + 1;
+                    else
+                        obj.bin_range_(1,i)    = last_bin_num;
+                    end
+                    
+                    obj.bin_edges_{i} = [];
+                    obj.read_pix_cache_{i}  = [];
+                end
+                if obj.bin_range_(2,i)< obj.bin_range_(1,i)
+                    obj.bin_range_(2,i) =obj.bin_range_(1,i);
+                    if obj.last_bin_completed_(i) && ...
+                            obj.bin_range_(2,i)<=obj.n_bins_ % we have not actually had
+                        % last bin the next bin data have never been placed
+                        % in cache
+                        obj.last_bin_completed_(i) = false;
+                    end
                 end
-                obj.bin_range_(1,i)    = last_bin_to_process+1;
                 
-                bin_filled(filled_bin_ind) = true;
+                
+                pix_tb(i,pix_ind) = arrayfun(@(x)(pic(:,left(x):right(x))),...
+                    bin_ind,'UniformOutput',false);
+                
+                %
+                next_bin_to_proc = min(next_bin_to_proc,obj.bin_range_(1,i));
             end
             
-            pix_tb = pix_tb(:,bin_filled); % accelerate combining by removing empty cells
             % combine pix from all files according to the bin
             pix_section = cat(2,pix_tb{:});
-            obj.first_bin_to_process_ = last_bin_to_process+1;
+            obj.first_bin_to_process_ = next_bin_to_proc;
             
-            if h_log_file
-                fprintf(h_log_file,' will save bins: [%d , %d]; ****** saving pixels: %d, pix left: %d\n',...
-                    first_bin_to_proc,last_bin_to_process,size(pix_section,2),npix_left);
-                fprintf(h_log_file,' cache contains: \n');
-                for j=1:n_files
-                    fprintf(h_log_file,'file %d; bin-range: [%d, %d] n full bins: %d\n',...
-                        j,obj.bin_range_(1,j),obj.bin_range_(2,j),numel(obj.filled_bin_ind_{j}));
-                end
-            end
+            %             if h_log
+            %                 fprintf(h_log,' will save bins: [%d , %d]; ****** saving pixels: %d, pix left: %d\n',...
+            %                     first_bin_to_proc,last_bin_to_process,size(pix_section,2),npix_left);
+            %                 fprintf(h_log,' cache contains: \n');
+            %                 for j=1:n_mess
+            %                     fprintf(h_log,'file %d; bin-range: [%d, %d] n full bins: %d\n',...
+            %                         j,obj.bin_range_(1,j),obj.bin_range_(2,j),numel(obj.filled_bin_ind_{j}));
+            %                 end
+            %             end
```

Comment:
Is this block needed any more?

Reply:
probably better to re-enable it, though hope one will never need to use it